narwhal/network/src/client.rs

```diff
@@ -0,0 +1,190 @@
+        Self::new(empty_peer_id())
+    }
+
+    pub fn set_worker_to_primary_local_handler(&self, handler: Arc<dyn WorkerToPrimary>) {
+        let mut inner = self.inner.write();
+        inner.worker_to_primary_handler = Some(handler);
+    }
+
+    pub fn set_primary_to_worker_local_handler(
+        &self,
+        worker_id: PeerId,
+        handler: Arc<dyn PrimaryToWorker>,
+    ) {
+        let mut inner = self.inner.write();
+        inner
+            .primary_to_own_worker_handler
+            .insert(worker_id, handler);
+    }
+
+    pub fn shutdown(&self) {
+        let mut inner = self.inner.write();
+        if inner.shutdown {
+            return;
+        }
+        inner.worker_to_primary_handler = None;
+        inner.primary_to_own_worker_handler = BTreeMap::new();
+        inner.shutdown = true;
+        let _ = self.shutdown_notify.notify();
+    }
+
+    async fn get_primary_to_worker_handler(
+        &self,
+        peer_id: PeerId,
+    ) -> Result<Arc<dyn PrimaryToWorker>, LocalClientError> {
+        for _ in 0..10 {
+            {
+                let inner = self.inner.read();
+                if inner.shutdown {
+                    return Err(LocalClientError::ShuttingDown);
+                }
+                if let Some(handler) = inner.primary_to_own_worker_handler.get(&peer_id) {
+                    return Ok(handler.clone());
+                }
+            }
+            sleep(Duration::from_millis(500)).await;
+        }
+        Err(LocalClientError::WorkerNotStarted(peer_id))
+    }
+
+    async fn get_worker_to_primary_handler(
+        &self,
+    ) -> Result<Arc<dyn WorkerToPrimary>, LocalClientError> {
+        for _ in 0..10 {
+            {
+                let inner = self.inner.read();
+                if inner.shutdown {
+                    return Err(LocalClientError::ShuttingDown);
+                }
+                if let Some(handler) = &inner.worker_to_primary_handler {
+                    debug!("Found primary {}", inner.primary_peer_id);
+                    return Ok(handler.clone());
+                }
+            }
+            sleep(Duration::from_millis(500)).await;
+        }
+        Err(LocalClientError::PrimaryNotStarted(
+            self.inner.read().primary_peer_id,
+        ))
+    }
+}
+
+// TODO: extract common logic for shutdown.
+
+#[async_trait]
+impl PrimaryToWorkerClient for NetworkClient {
+    async fn synchronize(
+        &self,
+        worker_peer: NetworkPublicKey,
+        request: WorkerSynchronizeMessage,
+    ) -> Result<(), LocalClientError> {
+        let c = self
+            .get_primary_to_worker_handler(PeerId(worker_peer.0.into()))
+            .await?;
+        select! {
+            resp = c.synchronize(Request::new(request)) => {
+                resp.map_err(|e| LocalClientError::Internal(format!("{e:?}")))?;
+                Ok(())
+            },
+            () = self.shutdown_notify.wait() => {
```

Comment:
this doesn't follow the existing shutdown convention and needs to be updated to match